php 8.1-8.2 compatibility

partikus opened the pull request:
This PR gonna be a successor of #5. I've realized that by the mistake I've touched @jack-worman PR. It should not happen any more. So I'm opening a new one.

More details:

1. all deps were added to `zf1s/phpunit` using `git subtree`, they have been readded because I've noticed changes which were not described in #5 
	```bash
    git subtree add --prefix deps/php-file-iterator https://github.com/sebastianbergmann/php-file-iterator.git 1.4 --squash -m "Merge sebastianbergmann/php-file-iterator:1.4 from https://github.com/sebastianbergmann/php-file-iterator/commit/730b01bc3e867237eaac355e06a36b85dd93a8b4\n\ngit subtree add --prefix deps/php-file-iterator https://github.com/sebastianbergmann/php-file-iterator.git 1.4 --squash -m \"\""
    git subtree add --prefix deps/php-text-template https://github.com/sebastianbergmann/php-text-template.git 1.2.1 --squash -m "Merge sebastianbergmann/php-text-template:1.2.1 from https://github.com/sebastianbergmann/php-text-template/commit/31f8b717e51d9a2afca6c9f046f5d69fc27c8686\n\ngit subtree add --prefix deps/php-text-template https://github.com/sebastianbergmann/php-text-template.git 1.2.1 --squash -m \"\""
    git subtree add --prefix deps/phpunit-mock-objects https://github.com/sebastianbergmann/phpunit-mock-objects.git 1.2.3 --squash -m "Merge sebastianbergmann/phpunit-mock-objects:1.2.3 from https://github.com/sebastianbergmann/phpunit-mock-objects/commit/5794e3c5c5ba0fb037b11d8151add2a07fa82875\n\ngit subtree add --prefix deps/phpunit-mock-objects https://github.com/sebastianbergmann/phpunit-mock-objects.git 1.2.3 --squash -m \"\""

	```
2. ~~I've switched extended version of autoloader to composer's repos using `path` types.~~
3. I've reverted `pear` repository.
4. more details: https://github.com/zf1s/phpunit/pull/9#issuecomment-1267467719

Many thanks to all contributors: @jack-worman @glensc  👏

partikus commented:
I'm not sure if merging subcomponents' history makes sense. That's longer discussion about pros & cons. I'm afraid some changes made by @jack-worman are not necessary, so thanks to that I've found them. 

Please vote 👎  or  👍  

@falkenhawk @marcing @glensc

partikus commented:
@falkenhawk @glensc ping

glensc commented:
The commit messages are messy here, what does this do? what is it different from:
1. #5 
2. #11

glensc commented:
maybe use git submodules instead of subtree? altho the immediate downside of submodule is inability to provide extra commits to it. but then again could fork them to repos in zf1s org.

glensc commented:
duh. can't checkout this with [gh](https://cli.github.com/) because same branch name used as in #5

```
➔ gh co 9
Already on 'Make_Php_8.1_Compatiable'
fatal: Not possible to fast-forward, aborting.
exit status 128
```

workaround: delete it

```
➔ g branch -D Make_Php_8.1_Compatiable
Deleted branch Make_Php_8.1_Compatiable (was 9a3de06d3).
➔ gh co 9
branch 'Make_Php_8.1_Compatiable' set up to track 'upstream/Make_Php_8.1_Compatiable'.
Switched to a new branch 'Make_Php_8.1_Compatiable'
```

partikus commented:
@glensc what's the messy? what could be improved?

I've rewritten the whole story from the beginning with another approach to keep the story trackable. I was not sure what was imported and modified with _"Inlined dependencies"._ 

I'm not happy with losing original commits but there was no way to keep them without bigger mess.

So there are the following changes:
- clear reference from where the changes come 
- keeping dependencies under single subfolder (less messy stuff in the project root)
- switching to custom composer's repositories to avoid downloading phpunit packages from github

I'm afraid we cannot use git submodules cause we have to apply some patches. So I'm afraid git submodules cannot help.

partikus commented:
@glensc yes, you can, here you go:

```bash
git checkout -b php81-compatibility-by-partikus upstream/Make_Php_8.1_Compatiable
```

Please let me know if you have any further questions.

glensc commented:
@partikus  please read my note till the end. I explained how I fixed it ;)
- https://github.com/zf1s/phpunit/pull/9#issuecomment-1267464599

after several reading and local checkout it's more understandable.

however, please see (and the whole pr):
- https://github.com/zf1s/phpunit/pull/11#issuecomment-1267469303

can you remove the big commit ede2a689d4268c84c5991992feaefbe63eb98d5f and cherry-pick the splitted-out versions from there:
- https://github.com/zf1s/phpunit/commit/63c40a6947f5448e7b3c44897c4870ab8161c4dc
- https://github.com/zf1s/phpunit/commit/94fc3efad831bed83b4e89b22b810f4cc4211f5d
- https://github.com/zf1s/phpunit/commit/7b326f41ada501a813805672ae332797914450b6

then I can close https://github.com/zf1s/phpunit/pull/11, it's only value was to resolve conflicts and make commits smaller and describe the changes.

partikus commented:
@glensc thank you for splitting the commit up, it works like a charm. I've cherry-picked changes. Looking forward to hearing your feedback.

glensc commented:
LGTM. I'm off to bed now.

glensc commented:
So. The composer thing, you have to declare repositories everywhere you use this phpunit fork? Maybe better adjust autoloader instead.

partikus commented:
@glensc your comment inspired me for doing more tests. It was not working properly without `replace` section. Of course, the repository entries did not work properly when the `zf1s/phpunit` was in the require section. So I've reapplied approach introduced in #5 using autoloader adjustment. 

Thanks!

falkenhawk commented:
@partikus I think the cleanest approach would be to fork the sub-dependencies and apply changes there.

partikus commented:
@glensc doing a fork of fork is an issue. It happens for us as well. I do agree with doing a mirror without forking. 
I'm not sure if we would have time for maintaining extra 3 packages including: (travis->github etc.), I'm not even sure if we should do that. Maybe mirroring the repo & applying patches is enough for us.

What do you think? @falkenhawk

falkenhawk commented:
Ok let's not create any additional forks. Copying the sub-deps over to this repo is fine, as long as the references to original repos are included in the "squashed git subtree" commits.

falkenhawk commented on README.md:
```diff
@@ -1,4 +1,6 @@
-# PHPUnit 3.7.x php 5.3-8.0 compatible
+# PHPUnit 3.7.x php 5.3-8.1 compatible
+
+> This project is not compatible with composer v2 which does not support pear repositories anymore. 
```
although this applies only to local dev/testing, as pear is included only in `require-dev`. `zf1s/phpunit` may still be installed with composer v2 from another project (namely `zf1s/zf1`) and it will install and work properly.

glensc replied:
If needed, could build composer repository with v1 that uses static files from github pages:
- https://github.com/eventum/composer

partikus commented:
@falkenhawk there is no way to modify the squashed commit' message. I've modified merge commits passing the url to appropriate commits. Github replaces the link with more human readable version.
Please let me know if it's fine for you.

<img width="484" alt="Made Php 8 1 Compatiable - sequel by partikus · Pull Request #9 · zf1s:phpunit 2022-10-07 22-03-12" src="https://user-images.githubusercontent.com/1732227/194644200-124f7a54-df54-4417-b8f4-a13083a7d21a.png">

falkenhawk commented:
@partikus thanks
Please adjust the PR description as some information stated there is no longer relevant.
And please adjust README for compatibility with php up to 8.2

falkenhawk reviewed (approved):
thank you all!!
